keycloak/keycloak-operator
# Intly 3226

## JIRA ID

INTLY-3226

## Additional Information

Adds a controller, reconciler and examples for Keycloak Realm reconciliation. The policy for realms is to only create but not manage them. Once created they are handed over to the Keycloak Admin user.

However some reconciliation is performed: if the credential secret for a user of the realm gets deleted, it will be recreated.

Realm users have to be created with the realm CR, see the example `realm_with_users.yaml`. Same goes for Clients, Identity providers etc.

Realms, like Keycloaks are multi namespace objects: they match to a Keycloak instance by using `spec.instanceSelector`. To test multi namespace support when running the operator locally you can use the `--scan-all` flag.

## Verification Steps

1. Run the operator from this branch
2. Install Keycloak by creating an instance of a Keycloak CR
3. Make sure that the Keycloak CR sets a label
4. Create an instance of a Keycloak Realm CR and make sure that it has an instance selector that matches the Keycloak label
5. Log into Keycloak and make sure the Realm has been created
6. If the realm had users, make sure they have been created and that a credential secret with the naming scheme `credential-<realm name>-<user name>-<keycloak namespace>` exists in the namespace of the realm

## Checklist:

## Additional Notes
